codegen: Improve array default handing

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45164 torch.fft: Two dimensional FFT functions
* **#45163 codegen: Improve array default handing**

Differential Revision: [D24132279](https://our.internmc.facebook.com/intern/diff/D24132279)

peterbell10 committed:
codegen: Improve array default handing

[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 commented on tools/codegen/api/cpp.py:
```diff
@@ -153,14 +153,23 @@ def returns_type(rs: Sequence[Return]) -> str:
+
+        return default_expr(d, t.elem)
+
+    if isinstance(t, ListType) and d.startswith('[') and d.endswith(']'):
+        return '{' + d[1:-1] + '}'
```
This is needed to handle `dim=[-2,-1]` in #45164 and more generally, to support arbitrary list defaults. Without it, it tries to literally put square-brackets into the C++ defaults.

ezyang replied:
I'd prefer it if we can make this fail noisily if someone accidentally, e.g., prefixes the default value with an extra space. Can we do something like, assert `d.startswith('[') and d.endswith(']')` when `isinstance(t, ListType)`? That would ensure that the person specified braces.

peterbell10 replied:
That check was there to allow sized lists with scalar defaults e.g. `int[2] padding=0`. So, I've put in an exception for unsized lists only.

mruberry commented:
Going to let @ezyang review this one. He's been working on codegen recently.

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

peterbell10 committed:
Update on "codegen: Improve array default handing"




[ghstack-poisoned]

mruberry commented:
Sorry to bother you @peterbell10, but would you rebase this please? Our internal tools are struggling to merge it properly.

peterbell10 committed:
Update on "codegen: Improve array default handing"


Differential Revision: [D24132279](https://our.internmc.facebook.com/intern/diff/D24132279)

[ghstack-poisoned]

peterbell10 commented:
@mruberry I fixed the merge conflict.
